### PR TITLE
Basic dex functionalities

### LINKED ISSUE #32: Implement function to place an limit order to the dex
We use the 0x protocol and will for the exchange REST api follow the standard relayer api https://github.com/0xProject/standard-relayer-api/blob/master/http/v0.md. 

All endpoints will be under api/v1/exchange

So this function needs to use the following REST endpoint:

POST api/v1/exchange/order

Submit a signed order to the relayer.
Payload

```
{
    "exchangeContractAddress": "0x12459c951127e0c374ff9105dda097662a027093",
    "maker": "0x9e56625509c2f60af937f23b7b532600390e8c8b",
    "taker": "0xa2b31dacf30a9c50ca473337c01d8a201ae33e32",
    "makerTokenAddress": "0x323b5d4c32345ced77393b3530b1eed0f346429d",
    "takerTokenAddress": "0xef7fff64389b814a946f3e92105513705ca6b990",
    "feeRecipient": "0xb046140686d052fff581f63f8136cce132e857da",
    "makerTokenAmount": "10000000000000000",
    "takerTokenAmount": "20000000000000000",
    "makerFee": "100000000000000",
    "takerFee": "200000000000000",
    "expirationUnixTimestampSec": "42",
    "salt": "67006738228878699843088602623665307406148487219438534730168799356281242528500",
    "ecSignature": {
        "v": 27,
        "r": "0x61a3ed31b43c8780e905a260a35faefcc527be7516aa11c0256729b5b351bc33",
        "s": "0x40349190569279751135161d22529dc25add4f6069af05be04cacbda2ace2254"
    }
}
```

also take a look at https://github.com/0xProject/0x.js/tree/development/packages/connect and what we can use.

### LINKED ISSUE #33: List the orderbook of the DEX
We use the 0x protocol and will for the exchange REST api follow the standard relayer api https://github.com/0xProject/standard-relayer-api/blob/master/http/v0.md.

All endpoints will be under api/v1/exchange

So this function needs to query the following REST endpoint:

GET api/v1/exchange/orderbook

Retrieves the orderbook for a given token pair.
Parameters

    baseTokenAddress [string]: address of token designated as the baseToken in the currency pair calculation of price (required)
    quoteTokenAddress [string]: address of token designated as the quoteToken in the currency pair calculation of price (required)

Response
```
{
    "bids": [
        {
            "exchangeContractAddress": "0x12459c951127e0c374ff9105dda097662a027093",
            "maker": "0x9e56625509c2f60af937f23b7b532600390e8c8b",
            "taker": "0xa2b31dacf30a9c50ca473337c01d8a201ae33e32",
            "makerTokenAddress": "0x323b5d4c32345ced77393b3530b1eed0f346429d",
            "takerTokenAddress": "0xef7fff64389b814a946f3e92105513705ca6b990",
            "feeRecipient": "0xb046140686d052fff581f63f8136cce132e857da",
            "makerTokenAmount": "10000000000000000",
            "takerTokenAmount": "20000000000000000",
            "makerFee": "100000000000000",
            "takerFee": "200000000000000",
            "expirationUnixTimestampSec": "42",
            "salt": "67006738228878699843088602623665307406148487219438534730168799356281242528500",
            "ecSignature": {
                "v": 27,
                "r": "0x61a3ed31b43c8780e905a260a35faefcc527be7516aa11c0256729b5b351bc33",
                "s": "0x40349190569279751135161d22529dc25add4f6069af05be04cacbda2ace2254"
            }
        },
        ...
    ],
    "asks": [
        {
            "exchangeContractAddress": "0x12459c951127e0c374ff9105dda097662a027093",
            "maker": "0x9e56625509c2f60af937f23b7b532600390e8c8b",
            "taker": "0xa2b31dacf30a9c50ca473337c01d8a201ae33e32",
            "makerTokenAddress": "0xef7fff64389b814a946f3e92105513705ca6b990",
            "takerTokenAddress": "0x323b5d4c32345ced77393b3530b1eed0f346429d",
            "feeRecipient": "0xb046140686d052fff581f63f8136cce132e857da",
            "makerTokenAmount": "22000000000000000",
            "takerTokenAmount": "10000000000000000",
            "makerFee": "100000000000000",
            "takerFee": "200000000000000",
            "expirationUnixTimestampSec": "632",
            "salt": "54515451557974875123697849345751275676157243756715784155226239582178",
            "ecSignature": {
                "v": 27,
                "r": "0x61a3ed31b43c8780e905a260a35faefcc527be7516aa11c0256729b5b351bc33",
                "s": "0x40349190569279751135161d22529dc25add4f6069af05be04cacbda2ace2254"
            }
        },
        ...
    ]
}
```
    bids - array of signed orders where takerTokenAddress is equal to baseTokenAddress
    asks - array of signed orders where makerTokenAddress is equal to baseTokenAddress

Bids will be sorted in descending order by price, and asks will be sorted in ascending order by price. Within the price sorted orders, the orders are further sorted first by total fees, then by expiration in ascending order.

Also take a look at https://github.com/0xProject/0x.js/tree/development/packages/connect and what we can use

### LINKED ISSUE #34: Function to take an limit order from the DEX
With this function a limit order found in the orderbook can be taken. 
Because the exchange happens onchain, this functions needs to create an ethereum transaction which calls the following function:

```
/// @dev Fills the input order.
/// @param orderAddresses Array of order's maker, taker, makerToken, takerToken, and feeRecipient.
/// @param orderValues Array of order's makerTokenAmount, takerTokenAmount, makerFee, takerFee, expirationTimestampInSec, and salt.
/// @param fillTakerTokenAmount Desired amount of takerToken to fill.
/// @param makerPath Path in the currency network of the maker token or [] if no currency network
/// @param takerPath Path in the currency network of the taker token or [] if no currency network
/// @param v ECDSA signature parameter v.
/// @param r ECDSA signature parameters r.
/// @param s ECDSA signature parameters s.
function fillOrderTrustlines(
        address[5] orderAddresses,
        uint[6] orderValues,
        uint fillTakerTokenAmount,
        address[] memory makerPath,
        address[] memory takerPath,
        uint8 v,
        bytes32 r,
        bytes32 s)
        public
        returns (uint filledTakerTokenAmount)
```

Also take a look at https://github.com/0xProject/0x.js/tree/development/packages/0x.js and what we can use. But be aware that this function is slightly different than the original fillOrder function of 0x.

### PR DESCRIPTION
**Pull request by dakingha69**
Closes #32, closes #33 and closes #34 
- list orderbook
- make order
- take order

### CODE REVIEW

File: src/User.ts [hunk 1]
```diff
@@ -72,19 +72,28 @@ export class User {
     })
   }
 
-  public signMsg (rawMsg: string): Promise<string> {
-    return new Promise<string>((resolve, reject) => {
+  public signMsg (msgHash: string): Promise<any> {
```
Comment (berndbohmeier) [inline src/User.ts]: Should be called signMsgHash if it takes the Hash

File: tests/e2e/Exchange.test.ts [hunk 2]
```diff
@@ -0,0 +1,291 @@
+import 'mocha'
+import * as chai from 'chai'
+import * as chaiAsPromised from 'chai-as-promised'
+import { TLNetwork } from '../../src/TLNetwork'
+import { config, keystore1, user1, keystore2 } from '../Fixtures'
+
+chai.use(chaiAsPromised)
+
+describe('e2e', () => {
```
Comment (berndbohmeier) [inline tests/e2e/Exchange.test.ts]: Maybe more readable with await? We want to go using await, async more anyways, do we? But not nessasary for merge, we can open a new task for migrating to use async await more in the code

File: tests/e2e/Exchange.test.ts [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    })
+
+    describe('#getOrderbook()', () => {
+      it('should return orderbook', () => {
+        expect(tl1.exchange.getOrderbook(makerTokenAddress, takerTokenAddress))
+          .to.eventually.have.keys('asks', 'bids')
+      })
+    })
+
+    describe('#makeOrder()', () => {
```
Comment (berndbohmeier) [inline tests/e2e/Exchange.test.ts]: Improve that test a bit, it basically does not test anything

File: src/Exchange.ts [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+
+  public getEthWrappers (): Promise<any> {
+    return this.utils.fetchUrl('exchange/eth')
+  }
+
+  public async getOrderbook (
+    baseTokenAddress: string,
+    quoteTokenAddress: string,
+    { baseTokenDecimals, quoteTokenDecimals }: OrderbookOptions = {}
+  ): Promise<any> {
```
Comment (berndbohmeier) [inline src/Exchange.ts]: Can we here maybe use the actual return type?

File: src/Exchange.ts [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+    makerTokenAddress: string,
+    takerTokenAddress: string,
+    makerTokenValue: number | string,
+    takerTokenValue: number | string,
+    {
+      makerTokenDecimals,
+      takerTokenDecimals,
+      expirationUnixTimestampSec = 2524604400
+    }: ExchangeOptions = {}
+  ): Promise<any> {
```
Comment (berndbohmeier) [inline src/Exchange.ts]: Same here, can we here maybe use the actual return type?

File: src/Exchange.ts [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+            takerTokenAmount: utils.formatAmount(signedOrder.takerTokenAmount, takerDecimals)
+          }))
+      )
+    } catch (error) {
+      return Promise.reject(error)
+    }
+  }
+
+  /**
+   * Prepares an on-chain transaction for a fill order.
```
Comment (berndbohmeier) [inline src/Exchange.ts]: Documentation is outdated

File: src/Exchange.ts [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+   * @param orderAddresses Array of order's maker, taker, makerToken, takerToken, and feeRecipient.
+   * @param orderValues Array of order's makerTokenAmount, takerTokenAmount, makerFee, takerFee, expirationTimestampInSec, and salt.
+   * @param fillTakerTokenAmount Desired amount of takerToken to fill.
+   * @param makerPath Path in the currency network of the maker token or [] if no currency network
+   * @param takerPath Path in the currency network of the taker token or [] if no currency network
+   * @param v ECDSA signature parameter v.
+   * @param r ECDSA signature parameter r.
+   * @param s ECDSA signature parameter s.
+   */
+  public async prepTakeOrder (
```
Comment (berndbohmeier) [inline src/Exchange.ts]: Can we change the the signature of the function to

prepTakeOrder(signedOrder: singedOrder)

File: src/Exchange.ts [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+        ? await payment.getPath(
+          makerTokenAddress,
+          makerAddress,
+          user.address,
+          this.getPartialAmount(fillTakerTokenValue, takerTokenValue, makerTokenValue),
+          { decimals: makerTokenDecimals }
+        ) : {
+          path: [],
+          maxFees: 0,
+          estimatedGas: 21000,
```
Comment (berndbohmeier) [inline src/Exchange.ts]: This is not enough gas for a tokenTransfer

Comment (berndbohmeier) [reply src/Exchange.ts]: 40000 should be a good number for a normal token transfer

File: package.json [hunk 9]
```diff
@@ -47,8 +47,10 @@
   },
   "dependencies": {
     "bignumber.js": "^5.0.0",
+    "bn.js": "^4.11.8",
```
[the remark this answers is not in the document]
Comment (dakingha69) [reply package.json]: no actually not... removed it and replaced it with bignumber.js

File: src/Payment.ts [hunk 10]
```diff
@@ -51,7 +54,7 @@ export class Payment {
 
   public prepareEth (
     to: string,
-    value: number,
+    value: number | string,
```
[the remark this answers is not in the document]
Comment (dakingha69) [reply src/Payment.ts]: yes that was the idea behind this

File: src/TLNetwork.ts [hunk 11]
```diff
@@ -7,8 +7,11 @@ import { CurrencyNetwork } from './CurrencyNetwork'
 import { Contact } from './Contact'
 import { Utils } from './Utils'
 import { Event } from './Event'
+import { Exchange } from './Exchange'
 import { Messaging } from './Messaging'
 
+import { Observable } from 'rxjs/Observable'
```
[the remark this answers is not in the document]
Comment (dakingha69) [reply src/TLNetwork.ts]: removed it

File: src/User.ts (hunk 1 above)
Comment (dakingha69) [reply src/User.ts]: yes makes sense, changing it

File: tests/e2e/Exchange.test.ts (hunk 2 above)
Comment (dakingha69) [reply tests/e2e/Exchange.test.ts]: yes definitely wanted to do a big refactoring of the code and use `await` and `async` more. I will open an issue for this.

File: tests/e2e/Exchange.test.ts (hunk 3 above)
Comment (dakingha69) [reply tests/e2e/Exchange.test.ts]: done

File: src/Exchange.ts (hunk 4 above)
Comment (dakingha69) [reply src/Exchange.ts]: done

File: src/Exchange.ts (hunk 7 above)
Comment (dakingha69) [reply src/Exchange.ts]: yes wanted to do this anyways

File: src/Exchange.ts (hunk 8 above)
Comment (dakingha69) [reply src/Exchange.ts]: done

**Pull request merged**